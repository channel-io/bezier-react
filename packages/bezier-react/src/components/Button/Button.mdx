import {
  ArgsTable,
  Canvas,
  Story,
} from '@storybook/addon-docs'
import {
  Button,
} from '~/src/components/Button'

# Button

## Overview

버튼은 클릭을 통해 사용자에게 액션을 제공하는 컴포넌트입니다.

버튼은 크게 4가지 사이즈로 나뉩니다. 그 중 M이 가장 보편적이며, 공간 절약을 위해 S가 쓰입니다. List나 Input 등 액션이 필요할 때는 XS도 사용할 수 있습니다. 가입, 초대 UX 등 굉장히 강조해야 할 버튼의 경우, L, XL 사이즈를 사용할 수 있습니다.

<Canvas>
  <Story id="components-button--overview-cta" />
</Canvas>

<Canvas>
  <Story id="components-button--overview-floating" />
</Canvas>

## Usage

### Recipe: Call to action

- 한 화면 내에서 CTA 버튼은 가능한 한, 가장 중요한 버튼 1개만 primary로 지정합니다. 후순위로 colored button을 같이 넣어야 할 경우에는 secondary로 넣습니다. 그 외에는 secondary, tetiary button을 조합하여 넣습니다.
- 버튼 사이의 간격은 0 또는 6으로 합니다. 둘 이상의 버튼을 활용할 때는, `<ButtonGroup>` 컴포넌트를 활용하는 것을 권장합니다.

<Canvas>
  <Story id="components-button--usage-cta" />
</Canvas>


<Canvas>
  <Story id="components-button--usage-cta-2" />
</Canvas>

ButtonGroup에 대해 더 알아보려면 [스토리](/docs/components-buttongroup--playground)를 참조하세요.

### Recipe: Web links

- 보통의 경우, tertiary button을 적절한 icon과 함께 사용합니다.

<Canvas>
  <Story id="components-button--usage-web-links" />
</Canvas>

### Recipe: Composite Usage

- XS 크기의 버튼은 ListItem, SectionLabel 등 다른 컴포넌트에 함께 사용될 수 있습니다. 컴포넌트 좌우측에서 액션으로 사용됩니다.

<Story id="components-button--usage-composite" />

### Recipe: Button with various contents

- 기본적으로 `leftContent`, `rightContent` prop에 `BezierIcon`을 지정하여 좌우측에 아이콘이 들어가는 형태를 표현합니다.

<Canvas>
  <Story id="components-button--usage-various-contents-composite" />
</Canvas>

- 아이콘만 들어가는 버튼의 경우, `leftContent` prop만 사용합니다.

<Canvas>
  <Story id="components-button--usage-various-contents-icon-only" />
</Canvas>

- 아이콘 이외에 커스텀 컴포넌트가 들어가야 하는 상황에는, `leftContent`, `rightContent` prop에 `JSX.Element` 값을 지정할 수 있습니다.

<Canvas>
  <Story id="components-button--usage-various-contents-custom" />
</Canvas>

### Recipe: Button with asynchronous actions

- 비동기적인 액션을 처리하는 경우, 액션이 처리중인 상태를 사용자에게 노출해야 하고, 액션이 처리중이라면 버튼을 비활성화하여 액션이 중복 처리되는 것을 막는 것이 보통의 경우입니다. `loading`, `disabled` prop을 통해 이 usecase를 구현할 수 있습니다.

```tsx
const AsyncActionButton = () => {
  const [isFetching, setFetching] = useState(false)
  const handleClick = () => {
    setFetching(true)
    setTimeout(() => {
      setFetching(false)
    }, 1000)
  }
  return (
    <Button
      leftContent={PlayIcon}
      text="Click Me!"
      colorVariant="cobalt"
      styleVariant="primary"
      loading={isFetching}
      disabled={isFetching}
      onClick={handleClick}
    />
  )
}
```

<Canvas>
  <Story id="components-button--usage-async" />
</Canvas>

### Recipe: Button with dropdown

- 버튼을 눌러 dropdown, select 등 별도의 UI를 노출하는 경우, 해당 UI가 노출된 상태에서는 버튼이 "눌린 상태"를 유지하는 것이 자연스럽습니다.
- `active` prop을 통해 버튼이 "눌린 상태" 임을 표현할 수 있습니다.

```tsx
const OpenDropdownButton = () => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <div>
      <Button
        text="Select"
        rightContent={TriangleDownIcon}
        active={isOpen}
        colorVariant="monochrome-light"
        styleVariant="tertiary
        onClick={() => setIsOpen(true)}
      />

      { ... }
    </div>
  )
}
```

<Canvas>
  <Story id="components-button--usage-dropdown" />
</Canvas>

## Variants

### Color Variants

- `ButtonColorVariant` enum을 통해 정의되며, `colorVariant` prop으로 지정할 수 있습니다.

<Story id="components-button--variants-color" />

### Style Variants

- `ButtonStyleVariant` enum을 통해 정의되며, `styleVariant` prop으로 지정할 수 있습니다.

<Story  id="components-button--variants-style" />

### Size

- `ButtonSize` enum을 통해 정의되며, `size` prop으로 지정할 수 있습니다.

<Story id="components-button--variants-size" />

## API

<ArgsTable of={Button} />

## Version

- Available since v0.3.28
