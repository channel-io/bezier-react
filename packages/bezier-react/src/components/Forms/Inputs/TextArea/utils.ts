import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
