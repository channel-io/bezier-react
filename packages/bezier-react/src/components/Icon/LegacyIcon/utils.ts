/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
