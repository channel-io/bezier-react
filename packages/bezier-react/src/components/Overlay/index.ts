export { Overlay } from './Overlay'
export {
  OverlayPosition,
  type OverlayProps,
} from './Overlay.types'
