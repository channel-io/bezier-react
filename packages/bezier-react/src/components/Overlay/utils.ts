/* Internal dependencies */
import { css } from '~/src/foundation'
import {
  OverlayPosition,
  type ContainerRectAttr,
  type TargetRectAttr,
} from './Overlay.types'

const TOP_POSITION_OFFSET = 10

interface GetOverlayPositionArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr
  show: boolean
}

export function getOverlayPosition({
  containerRect,
  targetRect,
  show,
}: GetOverlayPositionArgs) {
  const { containerTop, containerLeft, scrollTop, scrollLeft } = containerRect
  const { targetTop, targetLeft, clientTop, clientLeft } = targetRect

  const top = targetTop - clientTop - containerTop + scrollTop
  const left = targetLeft - clientLeft - containerLeft + scrollLeft

  return {
    top: show ? top : top - TOP_POSITION_OFFSET,
    left,
  }
}

interface GetOverlayTranslatationArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr | null
  overlay: HTMLElement | null
  position: OverlayPosition
  marginX: number
  marginY: number
  keepInContainer: boolean
}

export function getOverlayTranslation({
  containerRect,
  targetRect,
  overlay,
  position,
  marginX,
  marginY,
  keepInContainer,
}: GetOverlayTranslatationArgs) {
  if (containerRect && targetRect && overlay) {
    const {
      containerWidth,
      containerHeight,
      containerTop,
      containerLeft,
    } = containerRect
    const { targetWidth, targetHeight, targetTop, targetLeft } = targetRect
    const { width: overlayWidth, height: overlayHeight } = overlay.getBoundingClientRect()

    let translateX = 0
    let translateY = 0

    // pre position
    switch (position) {
      case OverlayPosition.TopCenter:
      case OverlayPosition.TopLeft:
      case OverlayPosition.TopRight:
        translateY -= (overlayHeight + marginY)
        translateX += marginX
        break
      case OverlayPosition.RightCenter:
      case OverlayPosition.RightTop:
      case OverlayPosition.RightBottom:
        translateX += (targetWidth + marginX)
        translateY += marginY
        break
      case OverlayPosition.BottomCenter:
      case OverlayPosition.BottomLeft:
      case OverlayPosition.BottomRight:
        translateY += (targetHeight + marginY)
        translateX += marginX
        break
      case OverlayPosition.LeftCenter:
      case OverlayPosition.LeftTop:
      case OverlayPosition.LeftBottom:
        translateX -= (overlayWidth + marginX)
        translateY += marginY
        break
    }

    switch (position) {
      // post position
      case OverlayPosition.TopCenter:
      case OverlayPosition.BottomCenter:
        translateX -= ((overlayWidth / 2) - (targetWidth / 2))
        break
      case OverlayPosition.TopRight:
      case OverlayPosition.BottomRight:
        translateX -= (overlayWidth - targetWidth)
        break
      case OverlayPosition.RightCenter:
      case OverlayPosition.LeftCenter:
        translateY -= ((overlayHeight / 2) - (targetHeight / 2))
        break
      case OverlayPosition.RightBottom:
      case OverlayPosition.LeftBottom:
        translateY -= (overlayHeight - targetHeight)
        break

      // inner position
      case OverlayPosition.InnerLeftTop:
        translateX = marginX
        translateY = marginY
        break
      case OverlayPosition.InnerLeftBottom:
        translateX = marginX
        translateY = targetHeight - overlayHeight + marginY
        break
      case OverlayPosition.InnerRightTop:
        translateX = targetWidth - overlayWidth + marginX
        translateY = marginY
        break
      case OverlayPosition.InnerRightBottom:
        translateX = targetWidth - overlayWidth + marginX
        translateY = targetHeight - overlayHeight + marginY
        break
    }

    if (keepInContainer) {
      const isOverTop = targetTop + translateY < containerTop
      const isOverBottom = targetTop + translateY + overlayHeight > containerTop + containerHeight
      const isOverLeft = targetLeft + translateX < containerLeft
      const isOverRight = targetLeft + translateX + overlayWidth > containerLeft + containerWidth

      const topSpace = targetTop - containerTop
      const bottomSpace = (containerTop + containerHeight) - (targetTop + targetHeight)
      const leftSpace = targetLeft - containerLeft
      const rightSpace = (containerLeft + containerWidth) - (targetLeft + targetWidth)

      if (isOverTop) {
        translateY = topSpace > bottomSpace
          ? translateY
          : targetHeight - translateY - overlayHeight
      } else if (isOverBottom) {
        translateY = bottomSpace > topSpace
          ? translateY
          : targetHeight - translateY - overlayHeight
      }
      if (isOverLeft) {
        translateX = leftSpace > rightSpace
          ? translateX
          : targetWidth - translateX - overlayWidth
      } else if (isOverRight) {
        translateX = rightSpace > leftSpace
          ? translateX
          : targetWidth - translateX - overlayWidth
      }
    }

    return {
      translateX: Math.round(translateX),
      translateY: Math.round(translateY),
    }
  }
  return {
    translateX: 0,
    translateY: 0,
  }
}

interface GetOverlayStyleArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr | null
  overlay: HTMLElement | null
  position: OverlayPosition
  marginX: number
  marginY: number
  keepInContainer: boolean
  show: boolean
}

export function getOverlayStyle({
  containerRect,
  targetRect,
  overlay,
  position,
  marginX,
  marginY,
  keepInContainer,
  show,
}: GetOverlayStyleArgs) {
  if (containerRect && targetRect) {
    const { top, left } = getOverlayPosition({ containerRect, targetRect, show })
    const { translateX, translateY } = getOverlayTranslation({
      containerRect,
      targetRect,
      overlay,
      position,
      marginX,
      marginY,
      keepInContainer,
    })

    return css`
      top: ${top}px;
      left: ${left}px;
      transform: translateX(${translateX}px) translateY(${translateY}px);
    `
  }
  return css``
}
