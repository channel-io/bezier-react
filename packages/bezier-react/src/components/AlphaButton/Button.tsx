import React, { forwardRef } from 'react'

import { isBezierIcon } from '@channel.io/bezier-icons'
import classNames from 'classnames'

import {
  type ButtonProps,
  type ButtonSize,
} from '~/src/components/AlphaButton/Button.types'
import { AlphaSpinner } from '~/src/components/AlphaSpinner'
import { BaseButton } from '~/src/components/BaseButton'
import { Icon, type IconSize } from '~/src/components/Icon'
import { Text } from '~/src/components/Text'

import styles from './Button.module.scss'

function SideContent({
  size,
  content,
}: {
  size: IconSize
  content?: ButtonProps['prefixContent']
}) {
  return isBezierIcon(content) ? (
    <Icon
      source={content}
      size={size}
      className={styles.ButtonIcon}
    />
  ) : (
    content
  )
}

function getIconSize(size: ButtonSize) {
  return (
    {
      xs: 'xxs',
      s: 'xs',
      m: 's',
      l: 's',
      xl: 'm',
    } as const
  )[size]
}

function getTypography(size: ButtonSize) {
  return (
    {
      xs: '13',
      s: '13',
      m: '14',
      l: '15',
      xl: '18',
    } as const
  )[size]
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      as = BaseButton,
      text,
      prefixContent,
      suffixContent,
      color = 'blue',
      variant = 'primary',
      size = 'm',
      active,
      className,
      loading,
      ...rest
    },
    forwardedRef
  ) {
    const Comp = as as typeof BaseButton

    return (
      <Comp
        ref={forwardedRef}
        className={classNames(
          styles.Button,
          styles[`size-${size}`],
          styles[`variant-${variant}`],
          styles[`color-${color}`],
          active && styles.active,
          className
        )}
        {...rest}
      >
        <div
          className={classNames(
            styles.ButtonContent,
            loading && styles.loading
          )}
        >
          <SideContent
            size={getIconSize(size)}
            content={prefixContent}
          />

          {/* TODO: use AlphaText later, add typo */}
          <Text
            className={styles.ButtonText}
            typo={getTypography(size)}
            bold
          >
            {text}
          </Text>

          <SideContent
            size={getIconSize(size)}
            content={suffixContent}
          />
        </div>

        {loading && (
          <div
            className={classNames(
              styles.ButtonLoader,
              styles[`size-${getIconSize(size)}`]
            )}
          >
            <AlphaSpinner
              className={styles.Spinner}
              variant="on-overlay"
            />
          </div>
        )}
      </Comp>
    )
  }
)
