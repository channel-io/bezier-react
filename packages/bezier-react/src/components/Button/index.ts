import { Button } from './Button'
import {
  ButtonSize,
  ButtonColorVariant,
  ButtonStyleVariant,
} from './Button.types'
import type ButtonProps from './Button.types'

export type { ButtonProps }

export {
  Button,
  ButtonSize,
  ButtonColorVariant,
  ButtonStyleVariant,
}
