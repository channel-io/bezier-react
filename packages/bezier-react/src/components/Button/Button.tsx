import React, { forwardRef, useCallback } from 'react'

import { isBezierIcon } from '@channel.io/bezier-icons'
import classNames from 'classnames'

import { warn } from '~/src/utils/assert'

import { BaseButton } from '~/src/components/BaseButton'
import { Icon } from '~/src/components/Icon'
import { LegacyIcon, isIconName } from '~/src/components/LegacyIcon'
import { Spinner } from '~/src/components/Spinner'
import { Text } from '~/src/components/Text'

import type { ButtonProps, ButtonSize, SideContent } from './Button.types'

import styles from './Button.module.scss'

export const BUTTON_TEST_ID = 'bezier-button'

function getTypography(size: ButtonSize) {
  return (
    {
      xs: '13',
      s: '13',
      m: '14',
      l: '15',
      xl: '18',
    } as const
  )[size]
}

function getIconSize(size: ButtonSize) {
  return (
    {
      xs: 'xs',
      s: 'xs',
      m: 's',
      l: 's',
      xl: 'm',
    } as const
  )[size]
}

function getSpinnerSize(size: ButtonSize) {
  return (
    {
      xs: 'xs',
      s: 'xs',
      m: 's',
      l: 's',
      xl: 's',
    } as const
  )[size]
}

function ButtonSideContent({
  size,
  children,
}: {
  size: ButtonSize
  children: SideContent
}) {
  if (isIconName(children)) {
    warn(
      'Deprecation: IconName as a value for the leftContent property of a Button has been deprecated. Use the Icon of bezier-icons instead.'
    )
    return (
      <LegacyIcon
        className={styles.ButtonIcon}
        name={children}
        size={getIconSize(size)}
      />
    )
  }

  if (isBezierIcon(children)) {
    return (
      <Icon
        className={styles.ButtonIcon}
        source={children}
        size={getIconSize(size)}
      />
    )
  }

  return <>{children}</>
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      as = BaseButton,
      className,
      type = 'button',
      text,
      disabled = false,
      loading = false,
      active = false,
      size = 'm',
      styleVariant = 'primary',
      colorVariant = 'blue',
      leftContent,
      rightContent,
      onClick,
      ...rest
    },
    forwardedRef
  ) {
    const Comp = as as typeof BaseButton

    const handleClick = useCallback<React.MouseEventHandler<HTMLButtonElement>>(
      (event) => {
        if (!disabled) {
          onClick?.(event)
        }
      },
      [onClick, disabled]
    )

    return (
      <Comp
        // eslint-disable-next-line react/button-has-type
        type={type}
        ref={forwardedRef}
        className={classNames(
          styles.Button,
          styles[`size-${size}`],
          styles[`style-${styleVariant}`],
          styles[`color-${colorVariant}`],
          active && styles.active,
          className
        )}
        disabled={disabled}
        onClick={handleClick}
        data-testid={BUTTON_TEST_ID}
        data-bezier-component="Button"
        {...rest}
      >
        <div
          className={classNames(
            styles.ButtonContent,
            loading && styles.loading
          )}
        >
          <ButtonSideContent size={size}>{leftContent}</ButtonSideContent>

          {text && (
            <Text
              className={styles.ButtonText}
              typo={getTypography(size)}
              bold
            >
              {text}
            </Text>
          )}

          <ButtonSideContent size={size}>{rightContent}</ButtonSideContent>
        </div>

        {loading && (
          <div className={styles.ButtonLoader}>
            <Spinner size={getSpinnerSize(size)} />
          </div>
        )}
      </Comp>
    )
  }
)
