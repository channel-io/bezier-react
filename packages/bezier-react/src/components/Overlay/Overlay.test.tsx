import React from 'react'

import { fireEvent } from '@testing-library/dom'
import { getWindow } from 'ssr-window'

import { render } from '~/src/utils/test'

import { Button } from '~/src/components/Button'

import {
  CONTAINER_TEST_ID,
  ESCAPE_KEY,
  OVERLAY_TEST_ID,
  Overlay,
} from './Overlay'
import type {
  ContainerRectAttr,
  OverlayProps,
  TargetRectAttr,
} from './Overlay.types'
import { getOverlayTranslation } from './utils'

import styles from './Overlay.module.scss'

const RootOverlay: React.FC<OverlayProps> = ({ children, ...rests }) => (
  <div id="main">
    <Overlay {...rests}>
      { children }
    </Overlay>
  </div>
)

describe('Overlay', () => {
  let props: OverlayProps

  beforeEach(() => {
    props = {
      container: getWindow().document.body,
      show: true,
    }
  })

  describe('Position', () => {
    const overlay = {
      getBoundingClientRect: () => ({
        width: 400,
        height: 400,
      }),
    } as HTMLElement

    const targetRect: TargetRectAttr = {
      targetWidth: 100,
      targetHeight: 100,
      targetTop: 450,
      targetLeft: 450,
      clientTop: 0,
      clientLeft: 0,
    }

    const containerRect: ContainerRectAttr = {
      containerWidth: 1000,
      containerHeight: 1000,
      containerTop: 0,
      containerLeft: 0,
      scrollTop: 0,
      scrollLeft: 0,
    }

    describe('getOverlayTranslation', () => {
      it('Without any option', () => {
        const result = getOverlayTranslation({
          overlay: null,
          targetRect: null,
          position: 'bottom-center',
          marginX: 0,
          marginY: 0,
          keepInContainer: true,
          containerRect: {
            containerWidth: 0,
            containerHeight: 0,
            containerTop: 0,
            containerLeft: 0,
            scrollTop: 0,
            scrollLeft: 0,
          },
        })

        expect(result).toEqual({
          translateX: 0,
          translateY: 0,
        })
      })

      it('BottomLeft 일반적인 경우, targetHeight 만큼 이동.', () => {
        const result = getOverlayTranslation({
          overlay,
          targetRect,
          position: 'bottom-left',
          marginX: 0,
          marginY: 0,
          keepInContainer: true,
          containerRect,
        })

        expect(result).toEqual({
          translateX: 0,
          translateY: 100,
        })
      })

      it('BottomLeft 아래가 넘어가는 경우, -overlay.height 만큼 이동', () => {
        const overflowTarget: TargetRectAttr = {
          ...targetRect,
          targetTop: 950,
        }

        const result = getOverlayTranslation({
          overlay,
          targetRect: overflowTarget,
          position: 'bottom-left',
          marginX: 0,
          marginY: 0,
          keepInContainer: true,
          containerRect,
        })

        expect(result).toEqual({
          translateX: 0,
          translateY: -400,
        })
      })

      it('BottomLeft 아래가 넘어가지만 target 아래쪽 공간이 더 넓을 경우, 일반 상태와 같은 결과.', () => {
        const overflowTarget: TargetRectAttr = {
          ...targetRect,
          targetTop: 200,
        }

        const overflowContainer: ContainerRectAttr = {
          ...containerRect,
          containerHeight: 600,
        }

        /*
          containerHeight: 600,
          targetHeight: 100에
          targetTop: 200이므로,
          target의 아래쪽 공간은 300이다.
          300이 200보다 크므로, overlay는 아래쪽에 나타나야 함.
        */

        const result = getOverlayTranslation({
          overlay,
          targetRect: overflowTarget,
          position: 'bottom-left',
          marginX: 0,
          marginY: 0,
          keepInContainer: true,
          containerRect: overflowContainer,
        })

        expect(result).toEqual({
          translateX: 0,
          translateY: 100,
        })
      })
    })
  })

  describe('Props and Event', () => {
    const renderRootOverlay = (optionProps?: OverlayProps) => render(<RootOverlay {...props} {...optionProps} />)

    beforeEach(() => {
      props = {
        show: true,
        className: '',
        containerClassName: '',
        position: 'left-center',
        marginX: 0,
        marginY: 0,
        keepInContainer: false,
        withTransition: false,
        enableClickOutside: false,
        children: 'Test Overlay',
      }
    })

    describe('Props', () => {
      describe('show', () => {
        describe('is False', () => {
          it('container style', () => {
            const { container } = renderRootOverlay()
            // <main id="main" />
            expect(container.children.length).toBe(1)
          })
        })
      })

      describe('className', () => {
        it('is transferred', () => {
          const CLASSNAME = 'Test__Overlay'
          const { getByTestId } = renderRootOverlay({ className: CLASSNAME })
          const overlay = getByTestId(OVERLAY_TEST_ID)
          expect(overlay).toHaveClass(CLASSNAME)
        })
      })

      describe('style', () => {
        it('is transferred', () => {
          const STYLE: React.CSSProperties = {
            width: '100px',
          }
          const { getByTestId } = renderRootOverlay({ style: STYLE })
          const overlay = getByTestId(OVERLAY_TEST_ID)
          expect(overlay).toHaveStyle('width: 100px')
        })
      })

      describe('containerClassName', () => {
        it('is transferred', () => {
          const CLASSNAME = 'Test__Container'
          const { getByTestId } = renderRootOverlay({ containerClassName: CLASSNAME })
          const overlay = getByTestId(CONTAINER_TEST_ID)
          expect(overlay).toHaveClass(CLASSNAME)
        })
      })

      describe('containerStyle', () => {
        it('is transferred', () => {
          const STYLE: React.CSSProperties = {
            width: '100px',
          }
          const { getByTestId } = renderRootOverlay({ containerStyle: STYLE })
          const overlay = getByTestId(CONTAINER_TEST_ID)
          expect(overlay).toHaveStyle('width: 100px')
        })
      })

      describe('enableClickOutside', () => {
        document.onclick = jest.fn()
        const onHide = jest.fn()

        afterEach(jest.clearAllMocks)

        it('is True', () => {
          const { getByTestId } = renderRootOverlay({ enableClickOutside: true, onHide })
          const overlay = getByTestId(CONTAINER_TEST_ID)

          overlay.click()
          expect(document.onclick).toHaveBeenCalledTimes(1)
          expect(onHide).toHaveBeenCalledTimes(1)
          overlay.click()
          expect(document.onclick).toHaveBeenCalledTimes(2)
          expect(onHide).toHaveBeenCalledTimes(2)
        })

        it('is False - click is stopPropagation ', () => {
          const { getByTestId } = renderRootOverlay()
          const overlay = getByTestId(CONTAINER_TEST_ID)

          overlay.click()
          expect(document.onclick).toHaveBeenCalledTimes(0)
          expect(onHide).toHaveBeenCalledTimes(0)
          overlay.click()
          expect(document.onclick).toHaveBeenCalledTimes(0)
          expect(onHide).toHaveBeenCalledTimes(0)
        })
      })

      describe('withTransition', () => {
        it('is True', () => {
          const { getByTestId } = renderRootOverlay({ withTransition: true })
          const overlay = getByTestId(OVERLAY_TEST_ID)
          expect(overlay).toHaveClass(styles.transition)
        })
      })
    })

    describe('Event', () => {
      document.onkeydown = jest.fn()
      const onHide = jest.fn()

      afterEach(jest.clearAllMocks)

      describe('keydown', () => {
        it('is Triggered By Escape', () => {
          const { getByTestId } = renderRootOverlay({ withTransition: true, onHide })
          const overlay = getByTestId(OVERLAY_TEST_ID)
          fireEvent.keyDown(overlay, { key: ESCAPE_KEY })
          expect(document.onkeydown).toHaveBeenCalledTimes(1)
          expect(onHide).toHaveBeenCalledTimes(1)
          fireEvent.keyDown(overlay, { key: ESCAPE_KEY })
          expect(document.onkeydown).toHaveBeenCalledTimes(2)
          expect(onHide).toHaveBeenCalledTimes(2)
        })

        it('is not Triggered By All keys except Escape', () => {
          const { getByTestId } = renderRootOverlay({ withTransition: true, onHide })
          const overlay = getByTestId(OVERLAY_TEST_ID)
          fireEvent.keyDown(overlay, { key: 'Enter' })
          expect(document.onkeydown).toHaveBeenCalledTimes(1)
          expect(onHide).toHaveBeenCalledTimes(0)
          fireEvent.keyDown(overlay, { key: 'ArrowRight' })
          expect(document.onkeydown).toHaveBeenCalledTimes(2)
          expect(onHide).toHaveBeenCalledTimes(0)
          fireEvent.keyDown(overlay, { key: 'Z' })
          expect(document.onkeydown).toHaveBeenCalledTimes(3)
          expect(onHide).toHaveBeenCalledTimes(0)
        })
      })

      describe('click', () => {
        it('calls onHide when element outside the overlay is clicked', async () => {
          renderRootOverlay({ onHide })
          fireEvent.click(document.body)
          expect(onHide).toHaveBeenCalled()
        })

        it('does not call onHide when element inside the overlay is clicked', () => {
          const { getByRole } = renderRootOverlay({ children: (<Button text="button" />), onHide })
          const button = getByRole('button')
          fireEvent.click(button)
          expect(onHide).not.toHaveBeenCalled()
        })
      })
    })
  })
})
