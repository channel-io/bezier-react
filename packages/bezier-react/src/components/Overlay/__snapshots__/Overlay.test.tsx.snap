// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay test > Snapshot > 1`] = `
.c0 {
  position: absolute;
  z-index: 1000;
}

<div
  class="c0"
  data-testid="bezier-react-overlay"
>
  <div>
    test
  </div>
</div>
`;
